nixos/modules/programs/shadow.nix

```
@@ -1,128 +1,220 @@
-    decide on which ids to use for declaratively defined users without a static
-    id
-  - Systemd compile time options -Dsystem-uid-max= and -Dsystem-gid-max=, used
-    by systemd for features like ConditionUser=@system and systemd-sysusers
-  */
-  loginDefs =
-    ''
-      DEFAULT_HOME yes
-
-      SYS_UID_MIN  400
-      SYS_UID_MAX  999
-      UID_MIN      1000
-      UID_MAX      29999
-
-      SYS_GID_MIN  400
-      SYS_GID_MAX  999
-      GID_MIN      1000
-      GID_MAX      29999
-
-      TTYGROUP     tty
-      TTYPERM      0620
-
-      # Ensure privacy for newly created home directories.
-      UMASK        077
-
-      # Uncomment this and install chfn SUID to allow non-root
-      # users to change their account GECOS information.
-      # This should be made configurable.
-      #CHFN_RESTRICT frwh
-
-    '';
-
-  mkSetuidRoot = source:
-    { setuid = true;
-      owner = "root";
-      group = "root";
-      inherit source;
-    };
-
+  cfg = config.security.loginDefs;
+  settingsFormat = pkgs.formats.keyValue { };
+  # Create custom toKeyValue generator
+  # see https://man7.org/linux/man-pages/man5/login.defs.5.html for config specification
+  toKeyValue = generators.toKeyValue {
+    mkKeyValue = generators.mkKeyValueDefault { } " ";
+  };
+  configFile = pkgs.writeText "login.defs" (toKeyValue cfg.settings);
+  yesNoType = enum [ "yes" "no" ];
+  mkSetuidRoot = source: {
+    setuid = true;
+    owner = "root";
+    group = "root";
+    inherit source;
+  };
 in
-
 {
-
   ###### interface
 
   options = {
+    security.loginDefs = {
+      chfnRestrict = mkOption {
+        description = mdDoc ''
+          Use chfn SUID to allow non-root users to change their account GECOS information.
+        '';
+        type = nullOr str;
+        default = null;
+      };
+
+      settings = mkOption {
+        description = mdDoc ''
+          Config options for the /etc/login.defs file, that defines
+          the site-specific configuration for the shadow password suite.
+          See login.defs(5) man page for available options.
+        '';
+        type = submodule {
+          freeformType = settingsFormat.type;
+          /* There are three different sources for user/group id ranges, each of which gets
+             used by different programs:
+             - The login.defs file, used by the useradd, groupadd and newusers commands
+             - The update-users-groups.pl file, used by NixOS in the activation phase to
+               decide on which ids to use for declaratively defined users without a static
+               id
+             - Systemd compile time options -Dsystem-uid-max= and -Dsystem-gid-max=, used
+               by systemd for features like ConditionUser=@system and systemd-sysusers
+          */
+          options = {
+            DEFAULT_HOME = mkOption {
+              description = mdDoc
+                "Indicate if login is allowed if we can't cd to the home directory.";
```

```suggestion
              description = mdDoc "Indicate if login is allowed if we can't cd to the home directory.";
```
same for the other options